NullPointerException at io.openliberty.tools.gradle.tasks.AbstractFeatureTask.getInstalledFeatures

This was reported by Craig Muchinsky. I think it may be similar to an issue we just resolved in ci.maven (https://github.com/OpenLiberty/ci.maven/issues/915).

Here is the stack trace:

org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':wdp-connect-flight-liberty:installFeature'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:208)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$$Lambda$640/0000000000000000.accept(Unknown Source)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:263)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:206)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:187)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:114)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:409)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:399)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:242)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:150)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:94)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:372)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:359)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:352)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:338)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$$Lambda$593/0000000000000000.execute(Unknown Source)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: java.lang.NullPointerException: Cannot invoke method getServerFeatures() on null object
        at io.openliberty.tools.gradle.tasks.AbstractFeatureTask.getInstalledFeatures(AbstractFeatureTask.groovy:137)
        at io.openliberty.tools.gradle.tasks.InstallFeatureTask_Decorated.getInstalledFeatures(Unknown Source)
        at io.openliberty.tools.gradle.tasks.InstallFeatureTask_Decorated$getInstalledFeatures.callCurrent(Unknown Source)
        at io.openliberty.tools.gradle.tasks.InstallFeatureTask.installFeature(InstallFeatureTask.groovy:44)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:104)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:58)

--- a/src/main/groovy/io/openliberty/tools/gradle/tasks/AbstractFeatureTask.groovy
+++ b/src/main/groovy/io/openliberty/tools/gradle/tasks/AbstractFeatureTask.groovy
@@ -121,32 +121,35 @@ public class AbstractFeatureTask extends AbstractServerTask {
     }
 
     protected Set<String> getInstalledFeatures() throws PluginExecutionException {
-        def pluginListedFeatures = getPluginListedFeatures(false)
+        // If getInstallFeatureUtil returns null, then features are installed via ant 
         def pluginListedEsas = getPluginListedFeatures(true)
-
         InstallFeatureUtil util = getInstallFeatureUtil(pluginListedEsas)
+        if(installFeaturesFromAnt) {
+            Set<String> featuresInstalledFromAnt;
+            if(server.features.name != null) {
+                featuresInstalledFromAnt = new HashSet<String>(server.features.name);
+                return featuresInstalledFromAnt;
+            }
+            else {
+                featuresInstalledFromAnt = new HashSet<String>();
+                return featuresInstalledFromAnt;
+            }
+        }
 
+        
+        def pluginListedFeatures = getPluginListedFeatures(false)
         def dependencyFeatures = getDependencyFeatures()
-
         def serverFeatures = null;
 
-        if(util == null) {
-            // Features installed from ant
-            return new HashSet<String>(Arrays.asList(server.features.name.join(",")))
-        }
-        
-        else {
-            // if DevMode provides a server directory parameter use that for finding the server features
-            if (serverDirectoryParam != null) {
-                serverFeatures = util.getServerFeatures(new File(serverDirectoryParam))
-            } else if (getServerDir(project).exists()) {
-                serverFeatures = util.getServerFeatures(getServerDir(project))
-            }
-
-            Set<String> featuresToInstall = InstallFeatureUtil.combineToSet(pluginListedFeatures, dependencyFeatures, serverFeatures)
-            return featuresToInstall 
+        // if DevMode provides a server directory parameter use that for finding the server features
+        if (serverDirectoryParam != null) {
+            serverFeatures = util.getServerFeatures(new File(serverDirectoryParam))
+        } else if (getServerDir(project).exists()) {
+            serverFeatures = util.getServerFeatures(getServerDir(project))
         }
 
+        Set<String> featuresToInstall = InstallFeatureUtil.combineToSet(pluginListedFeatures, dependencyFeatures, serverFeatures)
+        return featuresToInstall 
 
     }
 

--- a/src/main/groovy/io/openliberty/tools/gradle/tasks/AbstractFeatureTask.groovy
+++ b/src/main/groovy/io/openliberty/tools/gradle/tasks/AbstractFeatureTask.groovy
@@ -130,18 +130,19 @@ public class AbstractFeatureTask extends AbstractServerTask {
 
         def serverFeatures = null;
 
-        // if DevMode provides a server directory parameter use that for finding the server features
-        if (serverDirectoryParam != null) {
-            serverFeatures = util.getServerFeatures(new File(serverDirectoryParam))
-        } else if (getServerDir(project).exists()) {
-            serverFeatures = util.getServerFeatures(getServerDir(project))
-        }
-
         if(util == null) {
             // Features installed from ant
             return new HashSet<String>(Arrays.asList(server.features.name.join(",")))
         }
+        
         else {
+            // if DevMode provides a server directory parameter use that for finding the server features
+            if (serverDirectoryParam != null) {
+                serverFeatures = util.getServerFeatures(new File(serverDirectoryParam))
+            } else if (getServerDir(project).exists()) {
+                serverFeatures = util.getServerFeatures(getServerDir(project))
+            }
+
             Set<String> featuresToInstall = InstallFeatureUtil.combineToSet(pluginListedFeatures, dependencyFeatures, serverFeatures)
             return featuresToInstall 
         }

--- a/src/main/groovy/io/openliberty/tools/gradle/tasks/InstallFeatureTask.groovy
+++ b/src/main/groovy/io/openliberty/tools/gradle/tasks/InstallFeatureTask.groovy
@@ -44,7 +44,7 @@ class InstallFeatureTask extends AbstractFeatureTask {
         if(installFeaturesFromAnt) {
             installFeatureFromAnt();
         }
-        else if (util != null ) {
+        else {
             Set<String> featuresToInstall = getInstalledFeatures()
             util.installFeatures(server.features.acceptLicense, new ArrayList<String>(featuresToInstall))
         }

--- a/src/main/groovy/io/openliberty/tools/gradle/tasks/InstallFeatureTask.groovy
+++ b/src/main/groovy/io/openliberty/tools/gradle/tasks/InstallFeatureTask.groovy
@@ -41,7 +41,10 @@ class InstallFeatureTask extends AbstractFeatureTask {
         def pluginListedEsas = getPluginListedFeatures(true)
         InstallFeatureUtil util = getInstallFeatureUtil(pluginListedEsas)
 
-        if (util != null ) {
+        if(installFeaturesFromAnt) {
+            installFeatureFromAnt();
+        }
+        else if (util != null ) {
             Set<String> featuresToInstall = getInstalledFeatures()
             util.installFeatures(server.features.acceptLicense, new ArrayList<String>(featuresToInstall))
         }

--- a/src/main/groovy/io/openliberty/tools/gradle/tasks/InstallFeatureTask.groovy
+++ b/src/main/groovy/io/openliberty/tools/gradle/tasks/InstallFeatureTask.groovy
@@ -41,10 +41,10 @@ class InstallFeatureTask extends AbstractFeatureTask {
         def pluginListedEsas = getPluginListedFeatures(true)
         InstallFeatureUtil util = getInstallFeatureUtil(pluginListedEsas)
 
-        Set<String> featuresToInstall = getInstalledFeatures()
-
-
-        util.installFeatures(server.features.acceptLicense, new ArrayList<String>(featuresToInstall))
+        if (util != null ) {
+            Set<String> featuresToInstall = getInstalledFeatures()
+            util.installFeatures(server.features.acceptLicense, new ArrayList<String>(featuresToInstall))
+        }
     }
 
     private def buildAntParams() {

